### PR TITLE
Si 6196

### PR DESCRIPTION
**Pull request by rklaehn**
Added specialized implementation for filter

There is now an instrumented test checking for unnecessary object creations which the new implementation avoids, as well as a very simple correctness test under files/run

A benchmark is referenced in the JIRA issue, since AFAIK there is not yet a way to add a benchmark as a test.

### CODE REVIEW

**Comment by jsuereth** [comment]
Read my comment on your performance test.   I think, given performance findings, we need to be more cautious about polymorphism in collections.   I'd love to see numbers reflecting how performance degrades when using two different implementations of Set.

**Comment by rklaehn** [comment]
Hi Josh,

I have no idea what you mean by "what happens when the call-site becomes
megamorphic".

Regarding warmup: google caliper does appropriate warmup before each test.
It also runs every benchmark multiple times and only returns results when
the run time for each run are within a certain standard deviation of each
other. So the tests are representative of what you would get with a
long-running application where the JIT had time to do its job.

I think the instrumented test shows even better why the current
implementation is inferior. For a set with n elements of which k are
retained, the current implementation creates k unnecessary HashSet1[A]
objects, and it calls the hashcode method for each of the k retained
elements. In the given example with int entries hashcode is extremely
cheap. But for more complex elements unnecessary calls to hashcode can get
very expensive.

Regarding inlining: with all the stuff going on behind the scene (object
and array creations / invocations of hashcode and improve) I would be
surprised if inlining of the call site would make a big difference.

**Comment by rklaehn** [comment]
Another thing: should I adapt the benchmark and put it somewhere in the
scala repository? I saw that there are two directories called benchmark and
benchmarking, but there is no howto or anything, so I don't know how to get
started.

**Comment by rklaehn** [comment]
Is there still a chance of this going into 2.10, or should I close and resubmit the pull request to master?

File: src/library/scala/collection/immutable/HashSet.scala [hunk 1]
```diff
@@ -159,6 +169,16 @@ object HashSet extends ImmutableSetFactory[HashSet] {
           new HashSetCollision1(hash, ks1)
       } else this
 
+    protected override def filter0(p: A => Boolean) : HashSet[A] = {
+      val ks1 = ks.filter(p)
+      if (ks1.isEmpty)
+        null
```
Comment (jsuereth) [inline src/library/scala/collection/immutable/HashSet.scala:195]: Why don't you return HashSet.empty here and above instead of null?

Comment (rklaehn) [inline src/library/scala/collection/immutable/HashSet.scala:195]: I know it is ugly as hell. But it made a noticeable difference cases where
the resulting set is much smaller than the filtered set. Most pronounced in
the case filter(_=>false)  where the result is an empty set. Checking for
null in HashTrieSet.filter0 is quicker than calling x.isEmpty.

If this is too ugly for your taste I can take it out again and just live
with the performance penalty.

**Comment by rklaehn** [comment]
I guess this won't make it into 2.10 anymore, since it is not just a bugfix. I will open another pull request against master.

**Pull request closed without merging**